### PR TITLE
Implement unit tests for QRCodeOnboardingPayload

### PR DESCRIPTION
**Pull request by yufengwangca**
Implement unit tests for QRCodeOnboardingPayloadGenerator and QRCodeOnboardingPayloadParser, and add unit test to CI

### CODE REVIEW

File: src/controller/java/src/chip/onboardingpayload/QRCodeOnboardingPayloadParser.kt [hunk 1]
```diff
@@ -118,5 +85,38 @@ class QRCodeOnboardingPayloadParser(private val mBase38Representation: String) {
...
       return dest
     }
+
+    fun extractPayload(inString: String): String {
+      var chipSegment = ""
+      val delimiter = '%'
+      val startIndices = mutableListOf<Int>()
+      startIndices.add(0)
+
+      for (i in inString.indices) {
```
Comment (andy31415) [inline src/controller/java/src/chip/onboardingpayload/QRCodeOnboardingPayloadParser.kt:95]: I wonder if kotlin can be made shorter. maybe using `forEachIndexed` or even:

```kotlin
startIndices.addAll(
  inString
        .mapIndexed { i, c -> Pair(i,c)}
        .filter { v -> v.second == delimiter }
        .map { v -> v.first + 1 }
)
```

overall also the logic seems to be 'split by delimiters' so why are we not doing:

```kotlin
for (segment in inString.split('%')) {
  ...
}
```

File: src/controller/java/src/chip/onboardingpayload/QRCodeOnboardingPayloadParser.kt [hunk 2]
```diff
@@ -118,5 +85,38 @@ class QRCodeOnboardingPayloadParser(private val mBase38Representation: String) {
       index.addAndGet(numberOfBitsToRead)
       return dest
     }
+
+    fun extractPayload(inString: String): String {
```
Comment (andy31415) [inline src/controller/java/src/chip/onboardingpayload/QRCodeOnboardingPayloadParser.kt:89]: This seems to crash for `""` or `A%B%C` with `StringIndexOutOfBoundsException`

How about replacing this whole thing with:

```kotlin
fun extractPayload(inString: String): String {
    return inString
        .split('%')
        .filter { s -> s.startsWith(kQRCodePrefix)}
        .map { s -> s.substring(kQRCodePrefix.length)}
        .firstOrNull() ?: ""       
}
```

File: src/controller/java/src/chip/onboardingpayload/QRCodeOnboardingPayloadParser.kt (hunk 1 above)
Comment (andy31415) [reply src/controller/java/src/chip/onboardingpayload/QRCodeOnboardingPayloadParser.kt:95]: After this I tried to make the function cleaner, see above. Overall I believe kotlin code should not use indices/string operations as much and that may make things cleaner.

File: src/controller/java/tests/chip/onboardingpayload/QRCodeTest.kt [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+    assertEquals("", QRCodeOnboardingPayloadParser.extractPayload("MT:"))
+    assertEquals("", QRCodeOnboardingPayloadParser.extractPayload("H:"))
+    assertEquals("", QRCodeOnboardingPayloadParser.extractPayload("ASMT:"))
+    assertEquals("ABC", QRCodeOnboardingPayloadParser.extractPayload("Z%MT:ABC%"))
+    assertEquals("ABC", QRCodeOnboardingPayloadParser.extractPayload("%Z%MT:ABC%"))
+    assertEquals("ABC", QRCodeOnboardingPayloadParser.extractPayload("%Z%MT:ABC%DDD"))
+    assertEquals("ABC", QRCodeOnboardingPayloadParser.extractPayload("MT:ABC%DDD"))
+    assertEquals("ABC", QRCodeOnboardingPayloadParser.extractPayload("MT:ABC%"))
+    assertEquals("", QRCodeOnboardingPayloadParser.extractPayload("ABC"))
+    // TODO: support more match pattern
```
[the remark this answers is not in the document]
Comment (yufengwangca) [reply src/controller/java/tests/chip/onboardingpayload/QRCodeTest.kt:311]: Thanks for your comments, I combined your comments and here's the updated code:

```
fun extractPayload(inString: String?): String {
    return inString
        ?.split('%')
        ?.filter { s -> s.startsWith(kQRCodePrefix) }
        ?.firstOrNull()
        ?.substring(kQRCodePrefix.length)
        ?: ""
}
```

In this code, the safe navigation operator ?. is used to handle the case where inString is null. If inString is null, the entire chain of operations will short-circuit, and the function will return an empty string "".

If inString is not null, the operations will be executed as follows:

The string is split using the delimiter %.
The filter operation keeps only the substrings that start with the kQRCodePrefix.
The firstOrNull function retrieves the first filtered substring. If the list is empty, it returns null.
If the result from firstOrNull is not null, the substring function is called to remove the kQRCodePrefix.
If the result from substring is null, the elvis operator ?: is used to provide a default value of an empty string "".

**Review by andy31415 — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Comment by andy31415** [comment]
@yufengwangca need fixes for detekt static code analysis.

For constants I believe tests can have hardcoded numbers so we can update `kotlin-detect-config.yaml` to add an exclude for `**/Test*.kt` for magic numbers.

For the rest please check on a case by case basis.

**Pull request merged**